src/inverse/SIA2D/adjoint.jl

```diff
@@ -0,0 +1,415 @@
+    clamp_borders_dy_adjoint!(∂dSy, ∂Hlocy, ∂Cy, η₀, Δy, H, dSdy_edges)
+    ∇adj∂dSx = zero(S); ∇adj∂dSx[:,2:end - 1] .= diff_x_adjoint(∂dSx, Δx)
+    ∂C∂H_adj_x = ∇adj∂dSx + ∂Hlocx
+    ∇adj∂dSy = zero(S); ∇adj∂dSy[2:end - 1,:] .= diff_y_adjoint(∂dSy, Δy)
+    ∂C∂H_adj_y = ∇adj∂dSy + ∂Hlocy
+    ∂C∂H_adj = ∂C∂H_adj_x + ∂C∂H_adj_y
+
+    # Sum contributions of diffusivity and clipping
+    ∂H = ∂D∂H_adj + ∂C∂H_adj
+    ∂H .= ∂H.*(H.>0)
+
+    # Gradient wrt θ
+    ∂D∂θ = target.∂D∂θ(
+        H = H̄, ∇S = ∇S, θ = θ,
+        ice_model = SIA2D_model, ml_model = ml_model,
+        glacier = glacier, params = params
+    )
+    # Evaluate numerical integral for loss
+    @tullio ∂θ_v[k] := ∂D∂θ[i, j, k] * D_adjoint[i, j]
+    # Construct component vector
+    ∂θ = Vector2ComponentVector(∂θ_v, θ)
+
+    return ∂H, ∂θ
+end
+
+
+#######################################
+#####     Contiuous Adjoint      ######
+#######################################
+
+"""
+    VJP_λ_∂SIA∂H_continuous(
+        λ::Matrix{R},
+        H::Matrix{R},
+        simulation::SIM,
+        t::R;
+        batch_id::Union{Nothing, I} = nothing
+    )
+Implementation of the continuous adjoint of the SIA2D equation with respect to H.
+Given λ and H, it returns the VJP of λ^T * ∂(SIA2D)/∂H (H).
+
+Arguments:
+- `λ::Matrix{R}`: Adjoint state, also called output gradient in reverse-mode AD.
+- `H::Matrix{R}`: Ice thickness which corresponds to the input state of the SIA2D.
+- `simulation::SIM`: Simulation parameters.
+- `t::R`: Time value, not used as SIA2D is time independent.
+- `batch_id::Union{Nothing, I}`: Batch index.
+
+Returns:
+- `dλ::Matrix{R}`: Jacobian vector product, also called input gradient in reverse-mode AD.
+"""
+function VJP_λ_∂SIA∂H_continuous(
+    λ::Matrix{R},
+    H::Matrix{R},
+    θ,
+    simulation::SIM,
+    t::R;
+    batch_id::Union{Nothing, I} = nothing
+) where {R <: Real, I <: Integer, SIM <: Simulation}
+
+    if isnothing(batch_id)
+        SIA2D_model = simulation.model.iceflow
+        glacier = simulation.glaciers[SIA2D_model.glacier_idx[]]
+    else
+        SIA2D_model = simulation.model.iceflow[batch_id] # We pick the right iceflow model for this glacier
+        glacier = simulation.glaciers[batch_id]
+    end
+
+    # Retrieve models and parameters
+    params = simulation.parameters
+    ml_model = simulation.model.machine_learning
+    target = ml_model.target
+
+    # Retrieve parameters
+    B = glacier.B
+    Δx = glacier.Δx
+    Δy = glacier.Δy
+
+    # First, enforce values to be positive
+    map!(x -> ifelse(x > 0.0, x, 0.0), H, H)
+    # Update glacier surface altimetry
+    S = B .+ H
+
+    ### Computation of effective diffusivity
+    dSdx = Huginn.diff_x(S) ./ Δx
+    dSdy = Huginn.diff_y(S) ./ Δy
+    ∇Sx = Huginn.avg_y(dSdx)
+    ∇Sy = Huginn.avg_x(dSdy)
+
+    # Compute surface slope
+    ∇S = (∇Sx.^2 .+ ∇Sy.^2).^(1/2)
+
+    # Compute average ice thickness
+    H̄ = Huginn.avg(H)
```

Comment:
I realize we should probably export `avg` to have a cleaner interface.

Reply:
Yes, worth checking of this. If we include the `avg` function, we also probably want to include all the other similar utils from Huginn. I am not fully agains to leave it as it is tough.